### PR TITLE
migrate: support start desired version

### PR DESCRIPTION
**Pull request by yichengq**
from #2136

### CODE REVIEW

**Comment by yichengq** [comment]
@crawford Has made the change based on your comments.
@xiang90 ready to review.

File: main.go [hunk 1]
```diff
@@ -24,9 +24,15 @@
...
 
 import (
+	"os"
+
 	"github.com/coreos/etcd/etcdmain"
+	"github.com/coreos/etcd/migrate/starter"
 )
 
 func main() {
+	if os.Getenv("ETCD_START_DESIRED_VERSION") != "" {
```
Comment (crawford) [inline main.go]: Still the old name.

File: migrate/starter/starter.go [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+)
+
+type version string
+
+const (
+	internalV1      version = "1"
+	internalV2      version = "2"
+	internalUnknown version = "unknown"
+
+	defaultInternalV1etcdBinaryDir = "/var/lib/etcd/internal_versions/"
```
Comment (crawford) [inline migrate/starter/starter.go]: Still the old location.

**Comment by yichengq** [comment]
@crawford Sorry for that. I forget to push the latest version here. Check it again?

File: migrate/functional/member.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type Instance struct {
```
Comment (xiang90) [inline migrate/functional/member.go]: Proc

File: migrate/functional/upgrade_test.go [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+		os.Exit(1)
+	}
+	etcdctlBinPath = os.Getenv("ETCDCTL_BIN")
+
+	mustExist(v1BinPath)
+	mustExist(v2BinPath)
+	mustExist(etcdctlBinPath)
+}
+
+func TestStartSingleNewETCDUsingDefaultFlags(t *testing.T) {
```
Comment (xiang90) [inline migrate/functional/upgrade_test.go]: etcd never Etcd or ETCD

File: migrate/functional/upgrade_test.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+		if err != nil {
+			t.Fatalf("checkVersion error: %v", err)
+		}
+		if ver != "1" {
+			t.Errorf("internal version = %s, want %s", ver, "1")
+		}
+	}
+}
+
+func TestStartV2DesiredMultiV1ETCDUsingV1Flags(t *testing.T) {
```
Comment (xiang90) [inline migrate/functional/upgrade_test.go]: upgradeV1ClusterToV2Cluster

File: migrate/functional/upgrade_test.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+		if err != nil {
+			t.Fatalf("checkVersion error: %v", err)
+		}
+		if ver != "2" {
+			t.Errorf("internal version = %s, want %s", ver, "2")
+		}
+	}
+}
+
+func TestStartV1JoinedNewETCDUsingV1Flags(t *testing.T) {
```
Comment (xiang90) [inline migrate/functional/upgrade_test.go]: joinV1ClusterWithV1Flags

File: migrate/functional/upgrade_test.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+		if err != nil {
+			t.Fatalf("checkVersion error: %v", err)
+		}
+		if ver != "1" {
+			t.Errorf("internal version = %s, want %s", ver, "1")
+		}
+	}
+}
+
+func TestStartV1JoinedNewETCDUsingV1FlagsThroughDiscovery(t *testing.T) {
```
Comment (xiang90) [inline migrate/functional/upgrade_test.go]: joinV1ClusterViaDiscoveryWithV1Flags

File: migrate/functional/upgrade_test.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+	ver, err := checkInternalVersion(i.URL)
+	if err != nil {
+		t.Fatalf("checkVersion error: %v", err)
+	}
+	if ver != "2" {
+		t.Errorf("internal version = %s, want %s", ver, "2")
+	}
+}
+
+func TestStartMultiV1ETCDUsingV1Flags(t *testing.T) {
```
Comment (xiang90) [inline migrate/functional/upgrade_test.go]: TestRestartV1ClusterWithV1Flags

File: migrate/functional/upgrade_test.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+	ver, err := checkInternalVersion(i.URL)
+	if err != nil {
+		t.Fatalf("checkVersion error: %v", err)
+	}
+	if ver != "2" {
+		t.Errorf("internal version = %s, want %s", ver, "2")
+	}
+}
+
+func TestStartV2DesiredSingleV1ETCDWithSnapshotUsingV1Flags(t *testing.T) {
```
Comment (xiang90) [inline migrate/functional/upgrade_test.go]: TestUpgradeV1SnapshotedClusterToV2ClusterWithV1Flags

File: migrate/functional/upgrade_test.go [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+	ver, err := checkInternalVersion(i.URL)
+	if err != nil {
+		t.Fatalf("checkVersion error: %v", err)
+	}
+	if ver != "1" {
+		t.Errorf("internal version = %s, want %s", ver, "2")
+	}
+}
+
+func TestStartV2DesiredSingleV1ETCDUsingV1Flags(t *testing.T) {
```
Comment (xiang90) [inline migrate/functional/upgrade_test.go]: TestUpgradeV1ClusterToV2ClusterWithV1Flags

**Comment by yichengq** [comment]
@crawford @xiang90 Ready for the next round.

**Pull request merged**